src/aeso_aci.erl

```diff
@@ -0,0 +1,276 @@
+%%%-------------------------------------------------------------------
+%%% @author Robert Virding
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc
+%%%     ACI interface
+%%% @end
+%%% Created : 12 Dec 2017
+%%%-------------------------------------------------------------------
+
+-module(aeso_aci).
+
+-export([encode/1,encode/2,decode/1]).
+
+%% Define records for the various typed syntactic forms. These make
+%% the code easier but don't seem to exist elsewhere.
+
+-record(contract, {ann,con,decls}).
+-record(namespace, {ann,con,decls}).
+-record(letfun, {ann,id,args,type,body}).
+-record(type_def, {ann,id,vars,typedef}).
+
+-record(app_t, {ann,id,fields}).
+-record(tuple_t, {ann,args}).
+-record(record_t, {fields}).
+-record(field_t, {ann,id,type}).
+-record(alias_t, {type}).
+-record(variant_t, {cons}).
+-record(constr_t, {ann,con,args}).
+-record(fun_t, {ann,named,args,type}).
+
+-record(arg, {ann,id,type}).
+-record(id, {ann,name}).
+-record(con, {ann,name}).
+-record(qid, {ann,names}).
+-record(qcon, {ann,names}).
+-record(tvar, {ann,name}).
+
+%% encode(ContractString) -> {ok,JSON} | {error,String}.
+%% encode(ContractString, Options) -> {ok,JSON} | {error,String}.
+%%  Build a JSON structure with lists and tuples, not maps, as this
+%%  allows us to order the fields in the contructed JSON string.
+
+encode(ContractString) -> encode(ContractString, []).
+
+encode(ContractString, Options) when is_binary(ContractString) ->
+    encode(binary_to_list(ContractString), Options);
+encode(ContractString, Options) ->
+    try
+        Ast = parse(ContractString, Options),
+        %%io:format("~p\n", [Ast]),
+        %% aeso_ast:pp(Ast),
+        TypedAst = aeso_ast_infer_types:infer(Ast, Options),
+        %% io:format("~p\n", [TypedAst]),
+        %% aeso_ast:pp_typed(TypedAst),
+        %% We find and look at the last contract.
+        Contract = lists:last(TypedAst),
```

Comment:
This won't work if you are using types from other namespaces/contracts. For instance
```
contract OtherContract =
  datatype some_type = This | That
  function stuff : some_type => int

contract MyContract =
  function do_the_stuff(r : OtherContract, x : OtherContract.some_type) =
    r.stuff(x)
```
but I guess that's not really in scope for this PR.